File: javascript/easy/twoSumIII.js
/**
 * Initialize your data structure here.
 */
var TwoSum = function() {
    this.map = new Map();
};

/**
 * Add the number to an internal data structure..
 * @param {number} number
 * @return {void}
 */
TwoSum.prototype.add = function(number) {
    let map = this.map;
    if(map.get(number)) {
        map.set(number, map.get(number)+1);
    } else {
        map.set(number, 1);
    }
};

/**
 * Find if there exists any pair of numbers which sum is equal to the value.
 * @param {number} value
 * @return {boolean}
 */
TwoSum.prototype.find = function(target) {
    let map = this.map;
    for(let [key, value] of map) {
        if(((target - key) == key) ? (map.get(target-key) > 1) : (map.get(target-key))) {
            return true;
        }
    }
    return false;
};

/**
 * Your TwoSum object will be instantiated and called as such:
 * var obj = Object.create(TwoSum).createNew()
 * obj.add(number)
 * var param_2 = obj.find(value)
 */

File: ruby/easy/two_sum_iii.rb
class TwoSum

=begin
    Initialize your data structure here.
=end
    def initialize()
        @set = {}
    end


=begin
    Add the number to an internal data structure..
    :type number: Integer
    :rtype: Void
=end
    def add(number)
        if @set[number]
            @set[number] = @set[number]+1
        else
            @set[number] = 1
        end
    end


=begin
    Find if there exists any pair of numbers which sum is equal to the value.
    :type value: Integer
    :rtype: Boolean
=end
    def find(value)
        @set.each do |k,v|
            return true if ((value-k) == k) ? @set[value-k] > 1 : @set[value - k]
        end
        return false
    end


end

# Your TwoSum object will be instantiated and called as such:
# obj = TwoSum.new()
# obj.add(number)
# param_2 = obj.find(value)
